src/main/java/de/tum/in/www1/artemis/service/scheduled/ProgrammingExerciseScheduleService.java

```diff
@@ -123,11 +144,24 @@ private void scheduleExercise(ProgrammingExercise exercise) {
     }
 
     private void scheduleCourseExercise(ProgrammingExercise exercise) {
-        // TODO: there is small logic error here. When build and run test date is after the due date, the lock operation might be executed even if it is not necessary.
-        scheduleService.scheduleTask(exercise, ExerciseLifecycle.DUE, lockAllStudentRepositories(exercise));
-        scheduleService.scheduleTask(exercise, ExerciseLifecycle.BUILD_AND_TEST_AFTER_DUE_DATE, buildAndTestRunnableForExercise(exercise));
-        log.debug("Scheduled build and test for student submissions after due date for Programming Exercise \"" + exercise.getTitle() + "\" (#" + exercise.getId() + ") for "
-                + exercise.getBuildAndTestStudentSubmissionsAfterDueDate() + ".");
+        // For exercises with buildAndTestAfterDueDate or manual assessment
```

Comment:
@kloessst would that be better?
```suggestion
        // For any exercise that needsToBeScheduled (buildAndTestAfterDueDate and/or manual assessment)
```

Reply:
That's good. Just wasn't clear for me.